pkg/operators/ebpf/ebpf.go

```diff
@@ -0,0 +1,532 @@
+	ifaceName string
+
+	// map from ebpf variable name to ebpfVar struct
+	vars map[string]*ebpfVar
+
+	links             []link.Link
+	linksSnapshotters []*linkSnapshotter
+
+	containers map[string]*containercollection.Container
+
+	gadgetCtx operators.GadgetContext
+}
+
+func (i *ebpfInstance) loadSpec() error {
+	progReader := bytes.NewReader(i.program)
+	spec, err := ebpf.LoadCollectionSpecFromReader(progReader)
+	if err != nil {
+		return fmt.Errorf("loading spec: %w", err)
+	}
+	i.collectionSpec = spec
+	return nil
+}
+
+func (i *ebpfInstance) analyze() error {
+	prefixLookups := []populateEntry{
+		{
+			prefixFunc:   hasPrefix(tracerInfoPrefix),
+			validator:    i.validateGlobalConstVoidPtrVar,
+			populateFunc: i.populateTracer,
+		},
+		{
+			prefixFunc:   hasPrefix(snapshottersPrefix),
+			validator:    i.validateGlobalConstVoidPtrVar,
+			populateFunc: i.populateSnapshotter,
+		},
+		{
+			prefixFunc:   hasPrefix(paramPrefix),
+			validator:    i.validateGlobalConstVoidPtrVar,
+			populateFunc: i.populateParam,
+		},
+		// {
+		// 	prefixFunc:   hasPrefix(tracerMapPrefix),
+		// 	validator:    i.validateGlobalConstVoidPtrVar,
+		// 	populateFunc: i.populateMap,
+		// },
+		{
+			prefixFunc: func(s string) (string, bool) {
+				// Exceptions for backwards-compatibility
+				if s == gadgets.MntNsFilterMapName {
+					return gadgets.MntNsFilterMapName, true
+				}
+				if s == gadgets.MntNsFilterMapName {
+					return gadgets.MntNsFilterMapName, true
+				}
+				return "", false
+			},
+			populateFunc: i.populateMap,
+		},
+		{
+			prefixFunc: func(s string) (string, bool) {
+				// Exceptions for backwards-compatibility
+				if s == gadgets.FilterByMntNsName {
+					return gadgets.FilterByMntNsName, true
+				}
+				return hasPrefix(varPrefix)(s)
+			},
+			validator:    nil,
+			populateFunc: i.populateVar,
+		},
+	}
+
+	// Iterate over types and populate the gadget
+	it := i.collectionSpec.Types.Iterate()
+	for it.Next() {
+		for _, entry := range prefixLookups {
+			if prefix, ok := entry.prefixFunc(it.Type.TypeName()); ok {
+				if entry.validator != nil {
+					ok, err := entry.validator(it.Type, strings.TrimPrefix(it.Type.TypeName(), prefix))
+					if !ok {
+						if err != nil {
+							i.logger.Debugf("type %q error: %v", it.Type.TypeName(), err)
+						}
+						i.logger.Debugf("not matching")
+						continue
+					}
+				}
+				err := entry.populateFunc(it.Type, prefix) // strings.TrimPrefix(it.Type.TypeName(), prefix)
+				if err != nil {
+					return fmt.Errorf("handling type by prefix %q: %w", prefix, err)
+				}
+			}
+		}
+	}
+
+	// Iterate over programs
+	for name, program := range i.collectionSpec.Programs {
+		i.logger.Debugf("program %q", name)
+		i.logger.Debugf("> type       : %s", program.Type.String())
+		i.logger.Debugf("> attachType : %s", program.AttachType.String())
+		i.logger.Debugf("> sectionName: %s", program.SectionName)
+		i.logger.Debugf("> license    : %s", program.License)
```

Comment:
The logger is usually prefixing lines with a timestamp, so that could mess up the layout - but in general you're right, this is not a great solution. I'll think about it.